typing/typecore.ml

```diff
@@ -990,13 +968,110 @@ let rec has_literal_pattern p = match p.ppat_desc with
   | Ppat_or (p, q) ->
      has_literal_pattern p || has_literal_pattern q
 
-exception Need_backtrack
-
 let check_scope_escape loc env level ty =
   try Ctype.check_scope_escape env level ty
   with Unify trace ->
     raise(Error(loc, env, Pattern_type_clash(trace, None)))
 
+type pattern_checking_mode =
+  | Normal
+  (** We are checking user code. *)
+
+  | Counter_example of counter_example_checking_info
+(** In [Counter_example] mode, we are checking a counter-example candidate
+    produced by Parmatch. This is a syntactic pattern that represents a set of
+    values by using or-patterns (p_1 | ... | p_n) to enumerate all alternatives
+    in the counter-example search.
+
+    Parmatch does not use type information, so this pattern may exhibit two
+    issues:
+    - some parts of the pattern may be ill-typed due to GADTs.
+    - some wildcard patterns may not match any values: their type are empty.
+
+    The aim of [type_pat] in the [Counter_example] mode is to refine this
+    syntactic pattern into a well-typed pattern, and ensure that it matches
+    at least one concrete value.
+    - it filters ill-typed branches of or-patterns (see {!splitting_mode}.
+    - it tries to check that wildcard patterns are non-empty
+    (see {!explosion_fuel}).
+*)
+
+and counter_example_checking_info = {
+    explosion_fuel: int;
+    splitting_mode: splitting_mode;
+    constrs: (string, Types.constructor_description) Hashtbl.t;
+    labels: (string, Types.label_description) Hashtbl.t;
+  }
+(**
+    [explosion_fuel] controls the checking of wildcard patterns.
+    We eliminate potentially empty wildcard patterns by exploding them
```

Comment:
ditto above, it is not so much the or-pattern that is empty, as the whole pattern once we have chosen a branch in the or-pattern

Reply:
I added an extra comment on the splitting mode that talks (with examples) on the non-local nature of the search. I think it's enough to document it only there. I would claim that the subtlety you are pointing out is not due to wildcards in general, but in the or-patterns that some of those wildcards will explode into, so it's really related to or-patterns.